### PR TITLE
Emit `unhashable-member` when creating sets or dicts

### LINKED ISSUE #7034: Emit a message for unhashable dict keys
### Current problem

The following example leads to a `TypeError` because the key is unhashable.
pylint currently emits nothing for this:
```python
{[1, 2, 3]: "tomato"}
```

### Desired solution

Emit a message indicating the TypeError.

### Additional context

Noticed this while writing documentation for the `unhashable-dict-key` checker: https://github.com/PyCQA/pylint/pull/7033#issuecomment-1165864800.

Note the existing `unhashable-dict-key` checker emits a message when the key being looked up is unhashable; not the keys in the dict itself. I.e.:
```python
{"tomato": "red"}[["a", "b", "c"]]  # [unhashable-dict-key]
```

### LINKED ISSUE #7055: Emit a message for unhashable set items
### Current problem

The following example leads to a TypeError since the first item is not hashable:
```python
{[1, 2, 3], "tomato"}
```

### Desired solution

Emit a message indicating the TypeError.

### Additional context

Similar to: https://github.com/PyCQA/pylint/issues/7034 (unhashable dict keys).

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
Emit `unhashable-member` when creating sets or dicts, not just when accessing dicts.
Closes #7034
Closes #7055

### CODE REVIEW

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -1854,3 +1854,17 @@ def find_assigned_names_recursive(
...
+    try:
+        for inferred in node.infer():
+            if inferred is astroid.Uninferable:
+                return True
+            hash_fn = next(inferred.igetattr("__hash__"))
+            if getattr(hash_fn, "value", True) is not None:
+                return True
+    except astroid.InferenceError:
+        return True
+    return False
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1870]: Nitpick:
```suggestion
                return True
        return False
    except astroid.InferenceError:
        return True
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks great, I think there's some false positives detected by the primer though.

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -2015,6 +2016,20 @@ def visit_compare(self, node: nodes.Compare) -> None:
...
+        for k, _ in node.items:
+            if not is_hashable(k):
+                self.add_message("unhashable-dict-key", node=k, confidence=INFERENCE)
+
+    @only_required_for_messages("unhashable-dict-key")
+    def visit_set(self, node: nodes.Set) -> None:
+        for element in node.elts:
+            if not is_hashable(element):
+                self.add_message(
+                    "unhashable-dict-key", node=element, confidence=INFERENCE
```
Comment (mbyrnepr2) [inline pylint/checkers/typecheck.py:2030]: Do you think it could it deserve its own message; for example `unhashable-set-member`?

Comment (jacobtylerwalls) [reply pylint/checkers/typecheck.py:2030]: Or maybe we could rename to `unhashable-member` in a backwards-compatible way?

Comment (mbyrnepr2) [reply pylint/checkers/typecheck.py:2030]: That also sounds good.

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM, but we might want to do the name change in this merge request, right ?

File: doc/user_guide/checkers/features.rst [hunk 3]
```diff
@@ -1100,7 +1100,7 @@ Typecheck checker Messages
 :not-context-manager (E1129): *Context manager '%s' doesn't implement __enter__ and __exit__.*
   Used when an instance in a with statement doesn't implement the context
   manager protocol(__enter__/__exit__).
-:unhashable-dict-key (E1140): *Dict key is unhashable*
+:unhashable-member (E1140): *Member is unhashable*
   Emitted when a dict key is not hashable (i.e. doesn't define __hash__
```
Comment (mbyrnepr2) [inline doc/user_guide/checkers/features.rst:1104]:
```suggestion
  Emitted when a dict key or set member is not hashable (i.e. doesn't define __hash__
```

**Review by mbyrnepr2 — APPROVED** [review]

File: pylint/checkers/typecheck.py [hunk 4]
```diff
@@ -379,6 +374,13 @@ def _missing_member_hint(owner, attrname, distance_threshold, max_choices):
         "await-outside-async",
         "Emitted when await is used outside an async function.",
     ),
+    "E1143": (
+        "Member is unhashable",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:378]:
```suggestion
        "'%s' is unhashable and can't be used as a %s in a %s",
```

File: tests/functional/u/unhashable_member.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+unhashable-member:8:0:8:2::Member is unhashable:INFERENCE
```
Comment (Pierre-Sassoulas) [inline tests/functional/u/unhashable_member.txt:1]:
```suggestion
unhashable-member:8:0:8:2::'[1, 2, 3]' is unhashable and can't be used as a key in a dict:INFERENCE
```

File: pylint/checkers/typecheck.py [hunk 6]
```diff
@@ -2015,11 +2017,25 @@ def visit_compare(self, node: nodes.Compare) -> None:
         if op in {"in", "not in"}:
             self._check_membership_test(right)
 
+    @only_required_for_messages("unhashable-member")
+    def visit_dict(self, node: nodes.Dict) -> None:
+        for k, _ in node.items:
+            if not is_hashable(k):
+                self.add_message("unhashable-member", node=k, confidence=INFERENCE)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:2024]:
```suggestion
                self.add_message("unhashable-member", node=k, args=(k.as_string(), "key", "dict"), confidence=INFERENCE)
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
What do you think about a more detailed message ? Seems like we're pyling change on top of the initial MR, but if we add sets, then we need to rename the message, if we rename the message then we need to make it generic yet clear...

**Pull request merged**